### PR TITLE
[Hexagon] Makefile changes for hannk on Hexagon

### PR DESCRIPTION
**Pull request by aankit-ca**
Works on device.
Qurt crash on hexagon-sim.

@dsharlet @pranavb-ca Use this to test:
HL_TARGET=hexagon-32-qurt-hvx_128 make test-hexagon-device

### CODE REVIEW

File: apps/support/Makefile.inc [hunk 1]
```diff
@@ -103,6 +106,8 @@ CXX-host-metal ?= $(CXX)
 CXX-host-hvx_128 ?= $(CXX)
 CXX-host-hvx_64 ?= $(CXX)
 CXX-host-hvx ?= $(CXX)
+CXX-hexagon-32-qurt-hvx_128 ?= hexagon-clang++
+CXX-hexagon-32-qurt-hvx_64 ?= $(CXX-hexagon-32-qurt-hvx_128)
```
Comment (dsharletg) [inline apps/support/Makefile.inc:110]: I think we've been deprecating hvx_64, do we need this? (And same below)

File: apps/hannk/Makefile [hunk 2]
```diff
@@ -6,26 +6,64 @@ include ../support/Makefile.inc
...
 	$(BIN)/$(HL_TARGET)/compare_vs_tflite
 
 test: compare_vs_tflite
 	$(BIN)/$(HL_TARGET)/compare_vs_tflite test/*/* --benchmark 0
 
+test-hexagon-sim: $(BIN)/$(HL_TARGET)/$(BENCHMARK_OUT)
+	@mkdir -p $@
+	echo "Benchmarking $(TEST)..."
+	echo "$(HEXAGON_SDK_ROOT)/rtos/qurt/computev65/debugger/lnx64/qurt_model.so" > $(BIN)/$(HL_TARGET)/osam.cfg
+	echo "$(DEFAULT_HEXAGON_TOOLS_ROOT)/Tools/lib/iss/qtimer.so --csr_base=0xFC900000 --irq_p=1 --freq=19200000 --cnttid=1" > $(BIN)/$(HL_TARGET)/q6ss.cfg
```
Comment (dsharletg) [inline apps/hannk/Makefile:45]: How stable are these specific files/constants? Are they going to break often?

I have this comment on many things in this PR... It seems like a lot of the config stuff happening in this PR should be built into the tools themselves, so we don't need to bake so many things that could potentially break into our build?

**Review by dsharletg — COMMENTED** [review]
Thanks for this, very cool!

**Comment by steven-johnson** [comment]
This PR must also include equivalent changes to the CMake build script (which is the canonical build script for this app and the only one that is tested on the buildbots).

File: apps/support/Makefile.inc (hunk 1 above)
Comment (steven-johnson) [reply apps/support/Makefile.inc:110]: hvx_64 doesn't exist anymore. hvx_128 is just a synonym for hvx (which is now preferred). Remove hvx_64 and add hvx.

**Comment by steven-johnson** [comment]
> This PR must also include equivalent changes to the CMake build script (which is the canonical build script for this app and the only one that is tested on the buildbots).

(I'm happy to help make those changes once the rest of this PR is approved, BTW)

**Comment by steven-johnson** [comment]
Where does this PR stand?

**Comment by aankit-ca** [comment]
Hi Steven, I was OOO for a few days. I'm back now. I will update the PR soon with the requested changes.

File: apps/hannk/Makefile (hunk 2 above)
Comment (aankit-ca) [reply apps/hannk/Makefile:45]: The constants are stable for the same target (same qurt ‘variant’, like ‘computev65’). The configs are not present in the tools. The sdk examples also do the same thing to run.

File: apps/support/Makefile.inc [hunk 3]
```diff
@@ -103,6 +106,8 @@ CXX-host-metal ?= $(CXX)
 CXX-host-hvx_128 ?= $(CXX)
 CXX-host-hvx_64 ?= $(CXX)
 CXX-host-hvx ?= $(CXX)
+CXX-hexagon-32-qurt-hvx ?= hexagon-clang++
+CXX-hexagon-32-qurt-hvx_128 ?= $(CXX-hexagon-32-qurt-hvx)
```
Comment (dsharletg) [inline apps/support/Makefile.inc:110]: I think we should not add this `hvx_128` version of these flags either? I think the only non-deprecated thing is `hvx` right?

File: apps/hannk/Makefile [hunk 4]
```diff
@@ -6,26 +6,64 @@ include ../support/Makefile.inc
...
 CXXFLAGS += -Wno-unused-private-field -fno-exceptions -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wunused-variable -I$(MAKEFILE_DIR)
 
+BENCHMARK_OUT = benchmark
+ifeq (hexagon-32-qurt,$(findstring hexagon-32-qurt,$(HL_TARGET)))
+	# Building benchmark application as shared object instead of elf for
+	# hexagon target.
+	BENCHMARK_OUT := $(BENCHMARK_OUT).so
+	BENCHMARK_HEXAGON_FLAGS = -shared -fPIC -G0
+	HEXAGON_STUBS = $(BIN)/$(HL_TARGET)/stubs.o
+	TEST ?= test/misc/MUL.tflite
```
Comment (dsharletg) [inline apps/hannk/Makefile:16]: The benchmark app already runs every benchmark passed to it on the command line. Maybe we could push the whole folder, and run test/*/* like the non-Hexagon tests?

I think all of this logic should be in the test-hexagon-sim/test-hexagon-device target BTW, not stored in environment variables like TEST (that is way too generic).

Comment (aankit-ca) [reply apps/hannk/Makefile:16]: Hey Dillon, using `test/*/*` is not working for me. Now I realized that run_main_on_hexagon has a max-limit of 2048 characters.

Comment (aankit-ca) [reply apps/hannk/Makefile:16]: We can run all tests one at a time maybe?

Comment (dsharletg) [reply apps/hannk/Makefile:16]: I guess we have to. I think we should make all targets work this way though, not just Hexagon, so it is consistent.

File: apps/hannk/benchmark.cpp [hunk 5]
```diff
@@ -41,7 +41,7 @@ void run_benchmark(const std::string &filename, const InterpreterOptions &option
 
 }  // namespace hannk
 
-int main(int argc, char **argv) {
+__attribute__((visibility("default"))) int main(int argc, char **argv) {
```
Comment (steven-johnson) [inline apps/hannk/benchmark.cpp:44]: Why is this change needed?

File: apps/hannk/interpreter/stubs.cpp [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#include "stdio.h"
```
Comment (steven-johnson) [inline apps/hannk/interpreter/stubs.cpp:1]: Please run clang-format on this.

Comment (steven-johnson) [reply apps/hannk/interpreter/stubs.cpp:1]: Also please add comments about what it's for, etc.

File: src/CPlusPlusMangle.cpp [hunk 7]
```diff
@@ -533,7 +533,11 @@ std::string mangle_type(const Type &type, const Target &target, PrevPrefixes &pr
         case 16:
             return "s";
         case 32:
-            return "i";
+            if (target.arch == Target::Hexagon) {
```
Comment (steven-johnson) [inline src/CPlusPlusMangle.cpp:536]: Can you comment on what's going on here (and why we never noticed the issue before)?

File: apps/hannk/benchmark.cpp (hunk 5 above)
Comment (dsharletg) [reply apps/hannk/benchmark.cpp:44]: On Hexagon, this is getting compiled as a shared object, not an executable. So I'm guessing this makes main visible to dlsym.

File: apps/hannk/Makefile [hunk 8]
```diff
@@ -6,25 +6,52 @@ include ../support/Makefile.inc
...
 	$(BIN)/$(HL_TARGET)/compare_vs_tflite
 
 test: compare_vs_tflite
-	$(BIN)/$(HL_TARGET)/compare_vs_tflite test/*/* --benchmark 0
+	$(foreach test_model, $(shell ls -1 test/*/*), $(BIN)/$(HL_TARGET)/compare_vs_tflite $(test_model) --benchmark 0;)
+
+test-hexagon-sim: $(BIN)/$(HL_TARGET)/$(BENCHMARK_OUT)
+	@mkdir -p $@
+	echo "Benchmarking tests..."
+	echo "$(HEXAGON_SDK_ROOT)/rtos/qurt/computev65/debugger/lnx64/qurt_model.so" > $(BIN)/$(HL_TARGET)/osam.cfg
```
Comment (dsharletg) [inline apps/hannk/Makefile:42]: It looks like these constants could go in the run_hexagon_sim.sh script too?

File: apps/hannk/interpreter/stubs.cpp (hunk 6 above)
Comment (aankit-ca) [reply apps/hannk/interpreter/stubs.cpp]: Done

File: apps/hannk/benchmark.cpp (hunk 5 above)
Comment (aankit-ca) [reply apps/hannk/benchmark.cpp:47]: Right

File: apps/hannk/Makefile (hunk 8 above)
Comment (aankit-ca) [reply apps/hannk/Makefile:42]: Done

File: src/CPlusPlusMangle.cpp (hunk 7 above)
Comment (aankit-ca) [reply src/CPlusPlusMangle.cpp:536]: We never tested the name-mangling target feature for Hexagon and hence never saw the problem. I think the changes are needed because of Hexagon being a 32bit proc.

Comment (dsharletg) [reply src/CPlusPlusMangle.cpp:536]: If this is for 32-bit, shouldn't the condition be if (target.bits == 32) ?

But I think we test this code on other 32-bit targets, so I think this might be something else.

Comment (dsharletg) [reply src/CPlusPlusMangle.cpp:536]: @zvookin might know more

**Review by dsharletg — APPROVED** [review]
This looks pretty good to me, only some minor comments remaining. However, there are now conflicts in the Makefile.

File: apps/hannk/Makefile (hunk 8 above)
Comment (aankit-ca) [reply apps/hannk/Makefile]: Sorry forgot to do these.

File: src/CPlusPlusMangle.cpp (hunk 7 above)
Comment (aankit-ca) [reply src/CPlusPlusMangle.cpp:536]: @dsharletg  The headers have int32 defined as long and uint32 as unsigned long. That is why this change is needed on Hexagon.

**Review by dsharletg — APPROVED** [review]
Thanks for this, and for the iteration on the review!
